File: captchax/sdk/react-native/ios/CaptchaXModule.swift
import Foundation
import React

@objc(CaptchaXModule)
class CaptchaXModule: NSObject {

  private var baseUrl: String = "https://captchax.example.com"
  private var timeout: Int = 30000

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  func setBaseUrl(_ url: String) {
    baseUrl = url
  }

  @objc
  func setTimeout(_ timeout: Int) {
    self.timeout = timeout
  }

  @objc
  func getCaptcha(_ type: String,
                  resolver resolve: @escaping RCTPromiseResolveBlock,
                  rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let url = URL(string: "\(baseUrl)/api/v1/captcha/\(type)") else {
      reject("INVALID_URL", "Invalid base URL", nil)
      return
    }

    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.timeoutInterval = TimeInterval(timeout)

    let task = URLSession.shared.dataTask(with: request) { data, response, error in
      if let error = error {
        reject("NETWORK_ERROR", error.localizedDescription, error)
        return
      }

      guard let data = data else {
        reject("NO_DATA", "No data received", nil)
        return
      }

      do {
        let json = try JSONSerialization.jsonObject(with: data, options: [])
        resolve(json)
      } catch {
        reject("PARSE_ERROR", "Failed to parse response", error)
      }
    }

    task.resume()
  }

  @objc
  func verifyCaptcha(_ captchaId: String,
                     captchaType: String,
                     userResponse: NSDictionary,
                     track: NSArray,
                     resolver resolve: @escaping RCTPromiseResolveBlock,
                     rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let url = URL(string: "\(baseUrl)/api/v1/captcha/\(captchaType)/verify") else {
      reject("INVALID_URL", "Invalid base URL", nil)
      return
    }

    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.timeoutInterval = TimeInterval(timeout)

    let body: [String: Any] = [
      "captchaId": captchaId,
      "userResponse": userResponse as! [String: Any],
      "track": track as! [[String: Any]]
    ]

    do {
      request.httpBody = try JSONSerialization.data(withJSONObject: body, options: [])
    } catch {
      reject("SERIALIZATION_ERROR", "Failed to serialize request body", error)
      return
    }

    let task = URLSession.shared.dataTask(with: request) { data, response, error in
      if let error = error {
        reject("NETWORK_ERROR", error.localizedDescription, error)
        return
      }

      guard let data = data else {
        reject("NO_DATA", "No data received", nil)
        return
      }

      do {
        let json = try JSONSerialization.jsonObject(with: data, options: [])
        resolve(json)
      } catch {
        reject("PARSE_ERROR", "Failed to parse response", error)
      }
    }

    task.resume()
  }

  @objc
  func trackUserAction(_ x: Double,
                       y: Double,
                       timestamp: Double) -> NSDictionary {
    return [
      "x": x,
      "y": y,
      "timestamp": timestamp
    ]
  }
}

File: captchax/sdk/react-native/android/CaptchaXModule.kt
package com.captchax.sdk

import com.facebook.react.bridge.*
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.TimeUnit

class CaptchaXModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private var baseUrl: String = "https://captchax.example.com"
    private var timeout: Long = 30000

    override fun getName(): String {
        return "CaptchaXModule"
    }

    @ReactMethod
    fun setBaseUrl(url: String) {
        baseUrl = url
    }

    @ReactMethod
    fun setTimeout(timeoutMs: Int) {
        timeout = timeoutMs.toLong()
    }

    @ReactMethod
    fun getCaptcha(type: String, promise: Promise) {
        Thread {
            try {
                val url = URL("$baseUrl/api/v1/captcha/$type")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                connection.connectTimeout = timeout.toInt()
                connection.readTimeout = timeout.toInt()

                val responseCode = connection.responseCode
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    val response = connection.inputStream.bufferedReader().use { it.readText() }
                    promise.resolve(JSONObject(response).toString())
                } else {
                    promise.reject("NETWORK_ERROR", "HTTP error: $responseCode")
                }
            } catch (e: Exception) {
                promise.reject("NETWORK_ERROR", e.message, e)
            }
        }.start()
    }

    @ReactMethod
    fun verifyCaptcha(
        captchaId: String,
        captchaType: String,
        userResponse: ReadableMap,
        track: ReadableArray,
        promise: Promise
    ) {
        Thread {
            try {
                val url = URL("$baseUrl/api/v1/captcha/$captchaType/verify")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                connection.connectTimeout = timeout.toInt()
                connection.readTimeout = timeout.toInt()

                val body = JSONObject()
                body.put("captchaId", captchaId)
                body.put("userResponse", reactMapToJSON(userResponse))
                body.put("track", reactArrayToJSONArray(track))

                connection.outputStream.use { os ->
                    os.write(body.toString().toByteArray())
                }

                val responseCode = connection.responseCode
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    val response = connection.inputStream.bufferedReader().use { it.readText() }
                    promise.resolve(JSONObject(response).toString())
                } else {
                    promise.reject("VERIFICATION_ERROR", "HTTP error: $responseCode")
                }
            } catch (e: Exception) {
                promise.reject("VERIFICATION_ERROR", e.message, e)
            }
        }.start()
    }

    @ReactMethod
    fun trackUserAction(x: Double, y: Double, timestamp: Double): WritableMap {
        val map = Arguments.createMap()
        map.putDouble("x", x)
        map.putDouble("y", y)
        map.putDouble("timestamp", timestamp)
        return map
    }

    private fun reactMapToJSON(map: ReadableMap): JSONObject {
        val json = JSONObject()
        val iterator = map.keySetIterator()
        while (iterator.hasNextKey()) {
            val key = iterator.nextKey()
            when (map.getType(key)) {
                ReadableType.String -> json.put(key, map.getString(key))
                ReadableType.Number -> json.put(key, map.getDouble(key))
                ReadableType.Boolean -> json.put(key, map.getBoolean(key))
                ReadableType.Null -> json.put(key, null)
                else -> {}
            }
        }
        return json
    }

    private fun reactArrayToJSONArray(array: ReadableArray): org.json.JSONArray {
        val jsonArray = org.json.JSONArray()
        for (i in 0 until array.size()) {
            when (array.getType(i)) {
                ReadableType.Map -> jsonArray.put(reactMapToJSON(array.getMap(i)))
                ReadableType.String -> jsonArray.put(array.getString(i))
                ReadableType.Number -> jsonArray.put(array.getDouble(i))
                ReadableType.Boolean -> jsonArray.put(array.getBoolean(i))
                else -> {}
            }
        }
        return jsonArray
    }
}
